### PR TITLE
Master+libfdt 1.4.6 9

### PR DESCRIPTION
**Pull request by bryanodonoghue**
Following on from a discussions here: #2638 and here #2647 

This PR updates to libfdt 1.4.6-9.
In order to support OPTEE providing a DTB overlay we need to go to libfdt version 1.4.2 from the current 1.4.1.
1.4.7 is the latest version of libfdt but as is observed here https://github.com/ARM-software/arm-trusted-firmware/pull/1657 and pointed out by @etienne-lms to me subsequently, there is a serious performance issue with 1.4.7, so we have targeted 1.4.6-9 instead.

Two new libc files from newlib need to be pulled in to support

    strrchr
    strotoul

SPDX license headers have been added to the libc and libfdt imports.

Some very minor tidy-up work is required in libfdt - literally removing a couple of extra semicolons that were generating warnings for me. Other than that the upgraded 1.4.6-9 seems to work just fine.

A subsequent PR will do the actual work of the DTB overlay implementation but, I think its worthwhile getting the libfdt upgrade out sooner.

### CODE REVIEW

**Comment by bryanodonoghue** [comment]
@etienne-lms 

deckard@event-horizon:~/Development/dtc$ git remote -v
origin	https://git.kernel.org/pub/scm/utils/dtc/dtc.git (fetch)
origin	https://git.kernel.org/pub/scm/utils/dtc/dtc.git (push)
deckard@event-horizon:~/Development/dtc$ gsh
commit aadd0b65c987d21776354800fdf24ad35c0ffe68 (HEAD -> optee-base-v1.4.6-9)

Some minor fixups need to be done to that code
-fdt_set_hdr_(magic)
-fdt_set_hdr_(totalsize)
-fdt_set_hdr_(off_dt_struct)
-fdt_set_hdr_(off_dt_strings)
-fdt_set_hdr_(off_mem_rsvmap)
-fdt_set_hdr_(version)
-fdt_set_hdr_(last_comp_version)
-fdt_set_hdr_(boot_cpuid_phys)
-fdt_set_hdr_(size_dt_strings)
-fdt_set_hdr_(size_dt_struct)
+fdt_set_hdr_(magic);
+fdt_set_hdr_(totalsize);
+fdt_set_hdr_(off_dt_struct);
+fdt_set_hdr_(off_dt_strings);
+fdt_set_hdr_(off_mem_rsvmap);
+fdt_set_hdr_(version);
+fdt_set_hdr_(last_comp_version);
+fdt_set_hdr_(boot_cpuid_phys);

and then to remove a compile error
-       char data[];
+       char data[0];

Here is the branch without the fixes - the raw import from libfdt
https://github.com/bryanodonoghue/optee_os/tree/master%2Blibfdt-1.4.6-9-diff

File: lib/libutils/isoc/newlib/strrchr.c [hunk 1]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
```
[the remark this answers is not in the document]
Comment (bryanodonoghue) [reply lib/libutils/isoc/newlib/strrchr.c]: Sure thing. I hadn't realized you guys were on that version of newlib.

File: lib/libutils/isoc/newlib/strtoul.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+ * SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+#include "_ansi.h"
+#include <limits.h>
+#include <ctype.h>
+#include <stdlib.h>
+
+/*
```
[the remark this answers is not in the document]
Comment (bryanodonoghue) [reply lib/libutils/isoc/newlib/strtoul.c]: I'll push the raw import for the newlib stuff to this branch https://github.com/bryanodonoghue/optee_os/tree/master%2Blibfdt-1.4.6-9-diff

**Comment by bryanodonoghue** [comment]
@etienne-lms @jforissier 
OK.

So I've pulled in the strrchr and strtoul from newlib version 1.19.0 as we discussed.
To do that I've had to implement builtin functions for isalpha, isspace and isupper

I've pushed the original versions of strrchr, strtoul and libfdt to https://github.com/bryanodonoghue/optee_os/tree/master%2Blibfdt-1.4.6-9-diff

@etienne-lms that should answer your question on why the libfdt presented here differs from Rob Herring's commit, it is, minor changes to make gcc with OPTEE's compiler flags happy to build the import.

**Comment by etienne-lms** [comment]
@bryanodonoghue, Travis complains from checkpatch issues.
```
a1f3d41f libfdt: Move to version v1.4.6-9
WARNING: 'verison' may be misspelled - perhaps 'version'?
```

```
8b62b858 libutils: isoc: implement isalpha(), isspace() and isupper()
WARNING: Missing commit description - Add an appropriate one
```
Here, it seems your commit message shall contain at least 1 line of description between the message hear line and the s-o-b line.

With these 2 fixed: `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` . (and Travis should report a success)

**Comment by bryanodonoghue** [comment]
Eh well on the second one, that's the format used for
commit d815ab4721f65936a46204ddce62d4b922000ec8
libutils: isoc: implement tolower(), isdigit() and isxdigit()

So I assumed that was your preference here. I'll add text for that though and add your r-b, too.

Thanks

**Comment by etienne-lms** [comment]
@bryanodonoghue, you can add my `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` on change "libutils: Import strrchr from newlib" and "libutils: Import strtoul from newlib".

**Comment by bryanodonoghue** [comment]
Done.

Thank you for the review.

**Comment by jforissier** [comment]
Re. "libfdt: Move to version v1.4.6-9":

You should probably not include the upstream commit message. I find it confusing.

**Comment by etienne-lms** [comment]
Something like `Synchrosnized with libfdt from dtc-1.4.6-9gaadd0b65c987`?

**Comment by jforissier** [comment]
All good for me as it is now (1edde87), thanks.

**Pull request merged**